networkx/algorithms/lowest_common_ancestors.py

```diff
@@ -10,9 +10,125 @@
     "all_pairs_lowest_common_ancestor",
     "tree_all_pairs_lowest_common_ancestor",
     "lowest_common_ancestor",
+    "naive_lowest_common_ancestor",
+    "naive_all_pairs_lowest_common_ancestor",
 ]
 
 
+def get_a_lowest_common_ancestor(G, common_ancestors):
+    """Return a lowest common ancestor from the given set of common ancestors"""
+    # Start with an arbitrary node v from the set of common ancestors.  Follow
+    # arbitrary outgoing edges, remaining in the set of common ancestors, until
+    # reaching a node with no outgoing edge to another of the common ancestors.
+
+    v = next(iter(common_ancestors))
+    while True:
+        successor = None
+        for w in G.successors(v):
+            if w in common_ancestors:
+                successor = w
+                break
+        if successor is None:
+            return v
+        v = successor
+
+
+@not_implemented_for("undirected")
+@not_implemented_for("multigraph")
+def naive_all_pairs_lowest_common_ancestor(G, pairs=None):
+    """Return the lowest common ancestor of all pairs or the provided pairs
+
+    Parameters
+    ----------
+    G : NetworkX directed graph
+
+    pairs : iterable of pairs of nodes, optional (default: all pairs)
+        The pairs of nodes of interest.
+        If None, will find the LCA of all pairs of nodes.
+
+    Returns
+    -------
+    An iterator over ((node1, node2), lca) where (node1, node2) are
+    the pairs specified and lca is a lowest common ancestor of the pair.
+    Note that for the default of all pairs in G, we consider
+    unordered pairs, e.g. you will not get both (b, a) and (a, b).
+
+    Notes
+    -----
+    Only defined on non-null directed acyclic graphs.
+
+    See Also
+    --------
+    naive_lowest_common_ancestor
+    """
+    if not nx.is_directed_acyclic_graph(G):
+        raise nx.NetworkXError("LCA only defined on directed acyclic graphs.")
+    elif len(G) == 0:
+        raise nx.NetworkXPointlessConcept("LCA meaningless on null graphs.")
+    elif None in G:
+        raise nx.NetworkXError("None is not a valid node.")
+
+    ancestor_cache = {}
+
+    if pairs is None:
+        pairs = set()
+        for i, u in enumerate(G):
+            for j, v in enumerate(G):
+                if i <= j and (u, v) not in pairs:
+                    pairs.add((u, v))
+    else:
+        if type(pairs) != list:
+            pairs = list(pairs)
```

Comment:
Ahhh yes, this is happening because the iterator is being consumed in this checking loop. So the next loop over `pairs` ends up just being a loop over an empty iterator. Maybe we shouldn't even check this at all here. If `nx.ancestors` is used with a node which is not in the graph it gives a nice clean error, something like 

``` python
....
....

NetworkXError: The node 1100 is not in the digraph.
```

What do you think about removing this else condition here?

Reply:
Something like this https://github.com/networkx/networkx/pull/5736/files#r914724794